### ISSUE TITLE
Minecraft with OptiFine crashes immediately on JRE 10

### ISSUE THREAD

**Issue by Schyrsivochter** [sp614x/optifine#1739]
### Description
On my system, Minecraft with Optifine installed crashes immediately, before even writing something to the launcher’s log window, when JRE 10 is used (the default JRE on my system, which is also the launcher’s default setting).
This happens with both 1.12.2 and 1.13 (OF beta). 
Vanilla Minecraft 1.13 runs fine on JRE 10. All versions, vanilla and OF, run fine on JRE 8.

This is an excerpt from `/var/log/syslog` which gives us an insight to the cause of the crash (ak-ubuntu is my system’s name):
```
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: [0901/105721:INFO:LocalVersionProvider.cpp(81)] Looking for installed versions in /home/[CENSORED]/.minecraft/versions
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: [0901/105721:INFO:LauncherAppRenderer.cpp(110)] Javascript bridge extended.
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: Exception in thread "main" java.lang.ClassCastException: java.base/jdk.internal.loader.ClassLoaders$AppClassLoader cannot be cast to java.base/java.net.URLClassLoader
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: #011at net.minecraft.launchwrapper.Launch.<init>(Launch.java:34)
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: #011at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: [0901/105721:INFO:GameCallbacks.cpp(178)] launcher/launcher (main) info Process closed with exit code 256
Sep  1 10:57:21 ak-ubuntu minecraft-launcher.desktop[31001]: [0901/105721:INFO:ClientHandler.cpp(176)] Launcher is already open!
```
The corresponding part in `~/.minecraft/launcher_log.txt` is basically the same, only the exception message and stack trace are missing.

The cause seems to be that the structure of the JRE’s internal class loaders changed in version 9 or 10 so that a certain class loader which OF uses is no longer an instance of `java.net.URLClassLoader`. The code assumes that it is, though, and casts it unsafely.

### System and game details
**System:** Ubuntu 18.04.1 LTS (x86_64)
**Kernel:** 4.15.0-33-generic 
**JRE:** OpenJDK 10.0.2 build 10.0.2+13-Ubuntu-1ubuntu0.18.04.1
**Minecraft launcher:** 2.0.1005-dev-linux
**Minecraft game and OptiFine version:** 1.13-OptiFine_HD_U_E3_beta4, 1.12.2-OptiFine_HD_U_E2
**CPU:** Intel Core i5-7200U
**GPU:** Intel HD Graphics 620

**Comment by SetantaLP** [comment]
It's a problem with the launch wrapper not optifine itself, see #1376.

**Comment by Schyrsivochter** [comment]
Oh, shame … Would you mind explaining what the launch wrapper and legacy launcher are? I’m afraid I don’t understand Minecraft modding well enough.

**Comment by SetantaLP** [comment]
LegacyLauncher and launchwrapper are the same (but launchwrapper is the name of the jar that is used while the repository on github is named LegacyLauncher). The original purpose of this tool was to be able to launch minecraft versions that were released before the launcher exists, without adapting them to the way the launcher starts the game and handles data. To achieve this, the launchwrapper contains different tweak classes that are selected based on the version that should be started, but also allowed to set the tweak class via the command line. The fist way was used to start old minecraft version, while the second one was used by modders since it allowed to replace classes form the minecraft.jar with modified classes from a different jar-file when they are loaded by the jvm, instead of creating a modified minecraft.jar.

**Comment by sp614x** [comment]
Updated launchwrapper by Runemoro, should be compatible with Java 8,9,10:
[NewerJavaCompatibility.zip](https://github.com/sp614x/optifine/files/2342731/NewerJavaCompatibility.zip)

**Comment by Schyrsivochter** [comment]
I have a different OptiFine version, but I figured I could just replace `net.minecraft:launchwrapper:1.12` in the version’s JSON with `net.minecraft:launchwrapper:2.0` to use the patched launchwrapper. And it works! Thank you very much.
Where is that patched launchwrapper from?

**Comment by sp614x** [comment]
Tha patched launchwrapper was suggested by @skmedix on the OptiFine discord.
The file is from @Runemoro somehwhere on GitHub.

**Comment by gabriele2000** [comment]
> Tha patched launchwrapper was suggested by @skmedix on the OptiFine discord.
> The file is from @Runemoro somehwhere on GitHub.

Will you add this to the optifine installer or we have to wait until the "official" launchwrapper update?

**Comment by LightWayUp** [comment]
> Will you add this to the optifine installer or we have to wait until the "official" launchwrapper update?

I don't think Mojang is interested in updating LegacyLauncher anymore, now that Forge, the group that relied on it the most, is moving to their own launching system.

**Comment by HopefulFire** [comment]
Error details: File hash should be 6d7cc9b1a2a03e4da0a5269f00c61909bd1eb75c but was f50d6d9f4f13e2c457d5aad75ce441b969672e3f Permission denied
Filename on disk: 1.13.2-OptiFine_HD_U_E7.json
Path: /home/jade/.minecraft/versions/1.13.2-OptiFine_HD_U_E7/1.13.2-OptiFine_HD_U_E7.json
Exists: file

**Comment by HopefulFire** [comment]
You cannot change the .json file. So the launcher fix no longer works

**Comment by Schyrsivochter** [comment]
@HopefulFire what do you mean, you cannot change it? Do you mean that it reverts to the previous version when the launcher terminates? Make sure you make the changes while the launcher is not running.

**Comment by HopefulFire** [comment]
It seems to be working now, so I guess that must have been the problem. Thanks!

**Comment by sp614x** [comment]
Fixed installer to use launchwrapper-2.0
All new versions would use the new installer.
